design/package_structure/design.Rmd

```diff
@@ -0,0 +1,247 @@
+  preds <- predict(fit, type = "response", newdata = df)
+  ret <- matrix(preds, ncol = length(lvls), dimnames = list(row.names(fit$model), lvls))
+  if (ensure_unbias) {
+    ret <- ret - bias(fit, trt, strata)
+  }
+  ret
+}
+```
+
+```{r}
+
+pred <- predict_cf(fit, "trt")
+```
+
+
+## Robust variance
+
+```{r}
+adjust_pi <- function(pi_t) {
+  diag(pi_t) - pi_t %*% t(pi_t)
+}
+
+get_erb <- function(resi, strata, trt, pit, randomization) {
+  if (length(strata) == 0) {
+    return(0)
+  }
+  if (randomization %in% c("simple", "pocock-simon")) {
+    return(0)
+  }
+  # Calculate Omega Z under simple
+  omegaz_sr <- adjust_pi(pit)
+  idx <- split(seq_len(length(resi)), cbind(trt, strata))
+  resi_per_strata <- vapply(idx, function(ii) mean(resi[ii]), FUN.VALUE = 0)
+  # Calculate strata levels and proportions for
+  # the outer expectation
+
+  strata_props <- vapply(idx, length, FUN.VALUE = 0L)
+  strata_props <- strata_props / sum(strata_props)
+  # Estimate R(B) by first getting the conditional expectation
+  # vector for a particular strata (vector contains
+  # all treatment groups), then dividing by the pi_t
+
+  rb_z <- resi_per_strata / as.numeric(pit)
+  # Compute the R(B)[Omega_{SR} - Omega_{Z_i}]R(B) | Z_i
+  # for each Z_i
+  strata_levels <- length(resi_per_strata)
+  n_trt <- length(pit)
+  ind <- matrix(seq_len(strata_levels), byrow = TRUE, ncol = n_trt)
+  rb_z_sum <- lapply(
+    seq_len(nrow(ind)),
+    function(x) rb_z[ind[x, ]] %*% t(rb_z[ind[x, ]]) * sum(strata_props[ind[x, ]])
+  )
+  rb_z_sum <- Reduce(`+`, rb_z_sum)
+  rb_z_sum * omegaz_sr
+}
+
+vcov_robin <- function(fit, ...) {
+  UseMethod("vcov_robin")
+}
+
+are <- function(objs, cls) {
+  all(vapply(objs, is, class2 = cls, FUN.VALUE = TRUE))
+}
+
+vcov_robin.glm <- function(fit, trt, strat = NULL, preds = predict_cf(fit, trt, strat), sr_decompose = TRUE, randomization = "simple", ...) {
+  raw_data <- fit$data
+  resi <- residuals(fit, type = "response")
+  est <- colMeans(preds)
+  var_preds <- var(preds)
+  pit <- as.numeric(table(raw_data[[trt]]) / nrow(raw_data))
+  trt_lvls <- fit$xlevels[[trt]]
+
+  idx <- split(seq_len(nrow(raw_data)), raw_data[[trt]])
+  cov_ymu <- vapply(idx, function(is) stats::cov(fit$y[is], preds[is, ]), FUN.VALUE = rep(0, ncol(preds)))
+
+  vcov_sr <- vapply(idx, function(is) stats::var(fit$y[is]), FUN.VALUE = 0) / pit
+  if (sr_decompose) {
+    vcov_sr <- vcov_sr + (diag(var_preds) - 2 * diag(cov_ymu)) / pit
+  }
+  v <- diag(vcov_sr) + cov_ymu + t(cov_ymu) - var_preds
+  if (!randomization %in% c("simple", "pocock-simon") && length(strat) > 0) {
+    v <- v - get_erb(resi, raw_data[strat], raw_data[[trt]], pit, randomization)
+  }
+  ret <- v / length(resi)
+  dimnames(ret) <- list(trt_lvls, trt_lvls)
+  return(ret)
+}
+vcov_robin(fit, "trt", "z1")
+```
+
+
+## robincar_glm wrapper
+
+```{r}
+robincar_glm2 <- function(data, formula, trt, strata, car_scheme, ...) {
+  fit <- glm(formula = formula, data = data, ...)
```

Comment:
for negative binomial we need MASS::glm.nb

Reply:
the reason to have "unbiased prediction" is that we may encounter cases that the unbiased prediction itself is sufficient and the variance can be ontained from Jackknife/bootstrapping, especially when we want to combine the usage with some multiple imputation methods